diff --git a/configure b/configure
index 7be8fb1..998182e 100755
--- a/configure
+++ b/configure
@@ -1,14 +1,15 @@
 #! /bin/bash
 # This is not an autoconf generated configure
 #
+CC=gcc
 INCLUDE=${1:-"$PWD/include"}
 
 # Output file which is input to Makefile
 CONFIG=config.mk
 
 # Make a temp directory in build tree.
-TMPDIR=$(mktemp -d config.XXXXXX)
-trap 'status=$?; rm -rf $TMPDIR; exit $status' EXIT HUP INT QUIT TERM
+RANDOMDIR=$(mktemp -d config.XXXXXX)
+trap 'status=$?; rm -rf $RANDOMDIR; exit $status' EXIT HUP INT QUIT TERM
 
 check_prog()
 {
@@ -40,7 +41,7 @@ check_toolchain()
 
 check_atm()
 {
-    cat >$TMPDIR/atmtest.c <<EOF
+    cat >$RANDOMDIR/atmtest.c <<EOF
 #include <atm.h>
 int main(int argc, char **argv) {
 	struct atm_qos qos;
@@ -49,7 +50,7 @@ int main(int argc, char **argv) {
 }
 EOF
 
-    $CC -I$INCLUDE -o $TMPDIR/atmtest $TMPDIR/atmtest.c -latm >/dev/null 2>&1
+    $CC -I$INCLUDE -o $RANDOMDIR/atmtest $RANDOMDIR/atmtest.c -latm >/dev/null 2>&1
     if [ $? -eq 0 ]
     then
 	echo "TC_CONFIG_ATM:=y" >>$CONFIG
@@ -57,7 +58,7 @@ EOF
     else
 	echo no
     fi
-    rm -f $TMPDIR/atmtest.c $TMPDIR/atmtest
+    rm -f $RANDOMDIR/atmtest.c $RANDOMDIR/atmtest
 }
 
 check_xtables()
@@ -71,7 +72,7 @@ check_xtables()
 check_xt()
 {
     #check if we have xtables from iptables >= 1.4.5.
-    cat >$TMPDIR/ipttest.c <<EOF
+    cat >$RANDOMDIR/ipttest.c <<EOF
 #include <xtables.h>
 #include <linux/netfilter.h>
 static struct xtables_globals test_globals = {
@@ -90,13 +91,13 @@ int main(int argc, char **argv)
 }
 EOF
 
-    if $CC -I$INCLUDE $IPTC -o $TMPDIR/ipttest $TMPDIR/ipttest.c $IPTL \
+    if $CC -I$INCLUDE $IPTC -o $RANDOMDIR/ipttest $RANDOMDIR/ipttest.c $IPTL \
 	$(${PKG_CONFIG} xtables --cflags --libs) -ldl >/dev/null 2>&1
     then
 	echo "TC_CONFIG_XT:=y" >>$CONFIG
 	echo "using xtables"
     fi
-    rm -f $TMPDIR/ipttest.c $TMPDIR/ipttest
+    rm -f $RANDOMDIR/ipttest.c $RANDOMDIR/ipttest
 }
 
 check_xt_old()
@@ -108,7 +109,7 @@ check_xt_old()
     fi
 
     #check if we dont need our internal header ..
-    cat >$TMPDIR/ipttest.c <<EOF
+    cat >$RANDOMDIR/ipttest.c <<EOF
 #include <xtables.h>
 char *lib_dir;
 unsigned int global_option_offset = 0;
@@ -129,13 +130,13 @@ int main(int argc, char **argv) {
 
 EOF
 
-    $CC -I$INCLUDE $IPTC -o $TMPDIR/ipttest $TMPDIR/ipttest.c $IPTL -ldl >/dev/null 2>&1
+    $CC -I$INCLUDE $IPTC -o $RANDOMDIR/ipttest $RANDOMDIR/ipttest.c $IPTL -ldl >/dev/null 2>&1
     if [ $? -eq 0 ]
     then
 	echo "TC_CONFIG_XT_OLD:=y" >>$CONFIG
 	echo "using old xtables (no need for xt-internal.h)"
     fi
-    rm -f $TMPDIR/ipttest.c $TMPDIR/ipttest
+    rm -f $RANDOMDIR/ipttest.c $RANDOMDIR/ipttest
 }
 
 check_xt_old_internal_h()
@@ -147,7 +148,7 @@ check_xt_old_internal_h()
     fi
 
     #check if we need our own internal.h
-    cat >$TMPDIR/ipttest.c <<EOF
+    cat >$RANDOMDIR/ipttest.c <<EOF
 #include <xtables.h>
 #include "xt-internal.h"
 char *lib_dir;
@@ -168,14 +169,14 @@ int main(int argc, char **argv) {
 }
 
 EOF
-	$CC -I$INCLUDE $IPTC -o $TMPDIR/ipttest $TMPDIR/ipttest.c $IPTL -ldl >/dev/null 2>&1
+	$CC -I$INCLUDE $IPTC -o $RANDOMDIR/ipttest $RANDOMDIR/ipttest.c $IPTL -ldl >/dev/null 2>&1
 
 	if [ $? -eq 0 ]
 	then
 	    echo "using old xtables with xt-internal.h"
 	    echo "TC_CONFIG_XT_OLD_H:=y" >>$CONFIG
 	fi
-	rm -f $TMPDIR/ipttest.c $TMPDIR/ipttest
+	rm -f $RANDOMDIR/ipttest.c $RANDOMDIR/ipttest
 }
 
 check_ipt()
@@ -210,7 +211,7 @@ check_ipt_lib_dir()
 
 check_setns()
 {
-    cat >$TMPDIR/setnstest.c <<EOF
+    cat >$RANDOMDIR/setnstest.c <<EOF
 #include <sched.h>
 int main(int argc, char **argv)
 {
@@ -218,7 +219,7 @@ int main(int argc, char **argv)
 	return 0;
 }
 EOF
-    $CC -I$INCLUDE -o $TMPDIR/setnstest $TMPDIR/setnstest.c >/dev/null 2>&1
+    $CC -I$INCLUDE -D_GNU_SOURCE -o $RANDOMDIR/setnstest $RANDOMDIR/setnstest.c #>/dev/null 2>&1
     if [ $? -eq 0 ]
     then
 	echo "IP_CONFIG_SETNS:=y" >>$CONFIG
@@ -227,12 +228,12 @@ EOF
     else
 	echo "no"
     fi
-    rm -f $TMPDIR/setnstest.c $TMPDIR/setnstest
+    rm -f $RANDOMDIR/setnstest.c $RANDOMDIR/setnstest
 }
 
 check_ipset()
 {
-    cat >$TMPDIR/ipsettest.c <<EOF
+    cat >$RANDOMDIR/ipsettest.c <<EOF
 #include <linux/netfilter/ipset/ip_set.h>
 #ifndef IP_SET_INVALID
 #define IPSET_DIM_MAX 3
@@ -251,14 +252,14 @@ int main(void)
 #endif
 EOF
 
-    if $CC -I$INCLUDE -o $TMPDIR/ipsettest $TMPDIR/ipsettest.c >/dev/null 2>&1
+    if $CC -I$INCLUDE -o $RANDOMDIR/ipsettest $RANDOMDIR/ipsettest.c >/dev/null 2>&1
     then
 	echo "TC_CONFIG_IPSET:=y" >>$CONFIG
 	echo "yes"
     else
 	echo "no"
     fi
-    rm -f $TMPDIR/ipsettest.c $TMPDIR/ipsettest
+    rm -f $RANDOMDIR/ipsettest.c $RANDOMDIR/ipsettest
 }
 
 check_elf()
@@ -306,7 +307,7 @@ check_mnl()
 
 check_berkeley_db()
 {
-    cat >$TMPDIR/dbtest.c <<EOF
+    cat >$RANDOMDIR/dbtest.c <<EOF
 #include <fcntl.h>
 #include <stdlib.h>
 #include <db_185.h>
@@ -315,7 +316,7 @@ int main(int argc, char **argv) {
 	return 0;
 }
 EOF
-    $CC -I$INCLUDE -o $TMPDIR/dbtest $TMPDIR/dbtest.c -ldb >/dev/null 2>&1
+    $CC -I$INCLUDE -o $RANDOMDIR/dbtest $RANDOMDIR/dbtest.c -ldb >/dev/null 2>&1
     if [ $? -eq 0 ]
     then
 	echo "HAVE_BERKELEY_DB:=y" >>$CONFIG
@@ -323,7 +324,7 @@ EOF
     else
 	echo "no"
     fi
-    rm -f $TMPDIR/dbtest.c $TMPDIR/dbtest
+    rm -f $RANDOMDIR/dbtest.c $RANDOMDIR/dbtest
 }
 
 quiet_config()
diff --git a/ip/ipaddress.c b/ip/ipaddress.c
index 9797145..3922034 100644
--- a/ip/ipaddress.c
+++ b/ip/ipaddress.c
@@ -1589,7 +1589,7 @@ int print_addrinfo(const struct sockaddr_nl *who, struct nlmsghdr *n,
 	if (rta_tb[IFA_LOCAL]) {
 		print_color_string(PRINT_ANY,
 				   ifa_family_color(ifa->ifa_family),
-				   "local", "%s",
+				   "local", "%s ",
 				   format_host_rta(ifa->ifa_family,
 						   rta_tb[IFA_LOCAL]));
 		if (rta_tb[IFA_ADDRESS] &&
@@ -1604,7 +1604,7 @@ int print_addrinfo(const struct sockaddr_nl *who, struct nlmsghdr *n,
 					   format_host_rta(ifa->ifa_family,
 							   rta_tb[IFA_ADDRESS]));
 		}
-		print_int(PRINT_ANY, "prefixlen", "/%d", ifa->ifa_prefixlen);
+		print_int(PRINT_ANY, "prefixlen", "/%d ", ifa->ifa_prefixlen);
 	}
 
 	if (brief)
